relay/pkg/block/push.go

```diff
@@ -114,39 +141,102 @@ func (f *Forwarder) addHeaders(headers []*btc.Header) error {
 	return f.hostChain.AddHeaders(anchorHeader.Raw, packHeaders(headers))
 }
 
-func (f *Forwarder) updateBestHeader(
-	ctx context.Context,
-	newBestHeader *btc.Header,
-) error {
-	currentBestDigest, err := f.hostChain.GetBestKnownDigest()
+func (f *Forwarder) addHeadersWithRetarget(headers []*btc.Header) error {
+	epochStart := headers[0].Height - difficultyEpochDuration
+	epochEnd := epochStart + difficultyEpochDuration - 1
+
+	oldPeriodStartHeader, err := f.btcChain.GetHeaderByHeight(epochStart)
 	if err != nil {
-		return fmt.Errorf("could not get best known digest: [%v]", err)
+		return fmt.Errorf(
+			"could not get header by height [%v]: [%v]",
+			epochStart,
+			err,
+		)
 	}
 
-	currentBestHeader, err := f.btcChain.GetHeaderByDigest(currentBestDigest)
+	oldPeriodEndHeader, err := f.btcChain.GetHeaderByHeight(epochEnd)
 	if err != nil {
 		return fmt.Errorf(
-			"could not get current best header by digest: [%v]",
+			"could not get header by height [%v]: [%v]",
+			epochEnd,
 			err,
 		)
 	}
 
-	lastCommonAncestor, err := f.findLastCommonAncestor(
-		ctx,
-		newBestHeader,
-		currentBestHeader,
+	return f.hostChain.AddHeadersWithRetarget(
+		oldPeriodStartHeader.Raw,
+		oldPeriodEndHeader.Raw,
+		packHeaders(headers),
 	)
-	if err != nil {
-		return fmt.Errorf("could not find last common ancestor: [%v]", err)
-	}
+}
 
-	limit := newBestHeader.Height - lastCommonAncestor.Height + 1
+func (f *Forwarder) updateBestHeader(
+	ctx context.Context,
+	newBestHeader *btc.Header,
+) error {
+	totalAttempts := 30
 
-	return f.hostChain.MarkNewHeaviest(
-		lastCommonAncestor.Hash,
-		currentBestHeader.Raw,
-		newBestHeader.Raw,
-		big.NewInt(limit),
+	for attempt := 1; attempt <= totalAttempts; attempt++ {
+		logger.Infof(
+			"attempt [%v] to set header [%v] as new best",
+			attempt,
+			newBestHeader.Height,
+		)
+
+		currentBestDigest, err := f.hostChain.GetBestKnownDigest()
+		if err != nil {
+			return fmt.Errorf("could not get best known digest: [%v]", err)
+		}
+
+		currentBestHeader, err := f.btcChain.GetHeaderByDigest(
+			currentBestDigest,
+		)
+		if err != nil {
+			return fmt.Errorf(
+				"could not get current best header by digest: [%v]",
+				err,
+			)
+		}
+
+		lastCommonAncestor, err := f.findLastCommonAncestor(
+			ctx,
+			newBestHeader,
+			currentBestHeader,
+		)
+		if err != nil {
+			return fmt.Errorf("could not find last common ancestor: [%v]", err)
+		}
+
+		limit := big.NewInt(
+			newBestHeader.Height - lastCommonAncestor.Height + 1,
+		)
+
+		if willSucceed := f.hostChain.MarkNewHeaviestPreflight(
+			lastCommonAncestor.Hash,
+			currentBestHeader.Raw,
+			newBestHeader.Raw,
+			limit,
+		); willSucceed {
+			return f.hostChain.MarkNewHeaviest(
+				lastCommonAncestor.Hash,
+				currentBestHeader.Raw,
+				newBestHeader.Raw,
+				limit,
+			)
+		}
+
+		// wait a constant back-off time
+		select {
+		case <-time.After(10 * time.Second):
```

Comment:
Should 10 seconds value be extracted into constants in `forwarder.go`? Even if not, we should at least define it next to `totalAttempts` because only when they match each other, this loop makes sense. We now retry up to 5 minutes which makes sense. If we increased sleep time to e.g. 30 seconds, I'd argue 30 retries is too much.

Reply:
See https://github.com/keep-network/tbtc/commit/576e98d37e8d5c8c6d9b9d0a3640b40151e81dc3